### PR TITLE
Runtime differences for --unstable + type checking

### LINKED ISSUE #4334: discussion: Add --experimental flags for features that won't be stable for 1.0
These would be for all of the features which deno currently has but don't have a stable API yet.

An example of this are plugins (`--experimental-plugins`), which we probably can't stabilize before op crates (#2180) are added because they would probably share a lot of infrastructure in `deno_core`.

### PR DESCRIPTION
**Pull request by lucacasonato**
Based heavily on @kitsonk's work in #4639

Fixes #4334 

---

Co-authored-by: Kitson Kelly <me@kitsonkelly.com>

### CODE REVIEW

**Comment by lucacasonato** [comment]
If this looks ok up to now I'll move all of the unstable APIs into the right files.

cc @ry @bartlomieju

**Comment by kitsonk** [comment]
@lucacasonato coolness! 💯

File: cli/js/deno_unstable.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+export { openPlugin } from "./ops/plugins.ts";
+export { transpileOnly, compile, bundle } from "./compiler/api.ts";
+export { applySourceMap, formatDiagnostics } from "./ops/errors.ts";
+export { signal, signals, Signal, SignalStream } from "./signals.ts";
+export { setRaw } from "./ops/tty.ts";
+export { utimeSync, utime } from "./ops/fs/utime.ts";
+export { ShutdownMode, shutdown } from "./net.ts";
+export { listen, connect } from "./net_unstable.ts";
+
+export const __unstable = true;
```
Comment (bartlomieju) [inline cli/js/deno_unstable.ts:18]: This is cute, but not needed?

File: cli/js/lib.deno.unstable.d.ts [hunk 2]
```diff
@@ -0,0 +1,905 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+
+/// <reference no-default-lib="true" />
+
+declare namespace Deno {
+  const __unstable: true;
```
Comment (bartlomieju) [inline cli/js/lib.deno.unstable.d.ts:6]: Ditto

**Comment by lucacasonato** [comment]
How should I deal with the tests? Make the `unit_test_runner.ts` just run everything in unstable mode? Or should I split the unit tests into `unit_tests.ts` and `unit_tests_unstable.ts`? Should that be done in this PR or in a followup?

@ry @bartlomieju

**Comment by ry** [comment]
I'm a little concerned about how this is going to work with VS Code. The `deno types --unstable` means that type checking will not work for unstable things unless there's some configuration...

**Comment by lucacasonato** [comment]
> I'm a little concerned about how this is going to work with VS Code. The deno types --unstable means that type checking will not work for unstable things unless there's some configuration...

@axetroy's VS Code plugin already needs you to enable it via the `deno.enable` option in the VS Code settings. One could add a `deno.unstable` setting that toggles between the stable types or the unstable. I can open a PR to the vscode plugin that adds this functionality - it should be relatively simple.

**Comment by ry** [comment]
@lucacasonato Ok I'm good to land this now - could you please merge with master?

**Comment by lucacasonato** [comment]
@ry Sure - got a few questions though:

1. Is `Deno.permissions` actually unstable or is it just marked as such?
2. `Deno.Process` is marked as unstable. Should all of `Deno.Process` (and by extension `Deno.run` be unstable, or only `Deno.Process::kill`?
3. How should I move methods for certain classes to unstable (`Conn::closeWrite`)? I can't conditionally add them.

**Comment by ry** [comment]
1. Deno.permissions is actually unstable
2. Deno.Process is erroneously marked unstable - only the kill method is. 
3. Don't worry about solving this problem, just leave the kill method in the deno.ns.d.ts

**Pull request merged**